### ISSUE TITLE
NamedDataSource should implement EncodingAware

### ISSUE THREAD

**Issue by ben-manes** [bbottema/simple-java-mail#131]
When attaching a file, the content-type is inferred by `MimeBodyPart` by reading from the input stream. This occurs twice per attachment before the data is written out, resulting in 3 http calls. If the `DataSource` implements `javax.mail.EncodingAware` then this used if the result is non-null. When the file type is known, such as a pdf, this should avoid the extra hops. Unfortunately the given `DataSource` is decorated by `NamedDataSource`, so this optimization is not possible.

**Comment by bbottema** [comment]
Do you have a suggestion to how to solve this best?

**Comment by bbottema** [comment]
What about this?

```
class NamedDataSource implements DataSource, EncodingAware {
	
	(..)
	
	@Override
	public String getEncoding() {
		boolean dataSourceIsEncodingAware = this.dataSource != null && this.dataSource instanceof EncodingAware;
		return dataSourceIsEncodingAware ? ((EncodingAware) this.dataSource).getEncoding() : null;
	}
}
```

**Comment by ben-manes** [comment]
Yep, that is what I envisioned.

**Comment by bbottema** [comment]
Released in 5.0.2.
